plugins/experimental/ssl_session_reuse/src/openssl_utils.cc

```diff
@@ -0,0 +1,116 @@
+/** @file
+
+  openssl_utils.cc - Interaction with openssl constructs
+
+  @section license License
+
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+ */
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <string.h>
+#include <pthread.h>
+#include <errno.h>
+#include <math.h>
+#include <openssl/rand.h>
+#include <pthread.h>
+#include <netinet/in.h>
+#include <arpa/inet.h>
+#include <ts/ts.h>
+
+#include "ssl_utils.h"
+#include "Config.h"
+#include "session_process.h"
+#include "common.h"
+
+static int
+ssl_new_session(TSSslSessionID &sid)
+{
+  // Encode ID
+  std::string encoded_id;
+  int ret = encode_id(sid.bytes, sid.len, encoded_id);
+  if (ret < 0) {
+    TSError("encoded id failed.");
+    return 0;
+  }
+
+  int session_ret_len = SSL_SESSION_MAX_DER;
+  char session_data[SSL_SESSION_MAX_DER];
+  if (!TSSslSessionGetBuffer(&sid, session_data, &session_ret_len)) {
+    TSError("session data is too large. %d", session_ret_len);
+    return 0;
+  }
+
+  std::string encrypted_data;
+  ret = encrypt_session(session_data, session_ret_len, (unsigned char *)get_key_ptr(), get_key_length(), encrypted_data);
```

Comment:
Does encrypt_session() actually change the key, and if so, is it reentrant?

Reply:
No, it does not change the key.  encrypted_data is the output parameter.